content/blog/2024/2024-06-05-aws-s3-presigned-urls.md

```diff
@@ -0,0 +1,468 @@
+@EnableConfigurationProperties(AwsS3BucketProperties.class)
+public class StorageService {
+
+  private final S3Template s3Template;
+  private final AwsS3BucketProperties awsS3BucketProperties;
+
+  public URL generateViewablePresignedUrl(String objectKey) {
+    var bucketName = awsS3BucketProperties.getBucketName();
+    var urlValidity = awsS3BucketProperties.getPresignedUrlValidity();
+
+    return s3Template.createSignedGetURL(bucketName, objectKey, urlValidity);
+  }
+
+  public URL generateUploadablePresignedUrl(String objectKey) {
+    var bucketName = awsS3BucketProperties.getBucketName();
+    var urlValidity = awsS3BucketProperties.getPresignedUrlValidity();
+
+    return s3Template.createSignedPutURL(bucketName, objectKey, urlValidity);
+  }
+
+}
+```
+We have used the `S3Template` class provided by Spring Cloud AWS in our service layer which offers a high level abstraction over the `S3Presigner` class from the official AWS SDK.
+
+While it's possible to use the `S3Presigner` class directly, `S3Template` reduces boilerplate code and simplifies the generation of Presigned URLs by offering convenient, Spring-friendly methods.
+
+We also make use of our custom `AwsS3BucketProperties` class to reference the S3 bucket name and the Presigned URL validity duration defined in our `application.yaml` file.
+
+It's important to ensure that the **controller API endpoints that'll consumes the service class we've created, should be secured and preferably rate limited**. Before our application generates a Presigned URL, the requesting user's identity and authority should be validated to prevent unauthorized access.
+
+## Required IAM Permissions
+
+To have our service layer generate Presigned URLs correctly, the IAM user whose security credentials we have configured must have the necessary permissions of `s3:GetObject` and `s3:PutObject`.
+
+Here is what our policy should look like:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Action": [
+        "s3:GetObject",
+        "s3:PutObject"
+      ],
+      "Resource": "arn:aws:s3:::bucket-name/*"
+    }
+  ]
+}
+```
+
+The above IAM policy **conforms to the least privilege principle**, by granting only the necessary permissions required for our service layer to generate Presigned URLs. We also specify the bucket ARN in the `Resource` field, further limiting the scope of the IAM policy to work with a single bucket that is provisioned for our application.
+
+Additionally, we can further restrict access to our S3 bucket by allowing requests only from a specific IP address or IP address blocks where our client application is hosted.
+
+To achieve this, we can modify our IAM policy statement to include a condition that checks the source IP address of the request using <a href="https://aws.amazon.com/what-is/cidr/" target="_blank">CIDR notation</a>:
+
+```json
+"Condition": {
+  "IpAddress": {
+    "aws:SourceIp": "01.02.03.04/32"
```

Comment:
How does that work in practice if the client is a mobile application, for example. The client IP would change frequently. Can this configuration be updated per session by our backend?